### PR TITLE
nixos/restic: small enhancements

### PR DESCRIPTION
**Pull request by robryk**
###### Description of changes

This PR adds an assertion that detects configuration that will either not work or will have its part ignored and adds a way to specify ignored file patterns that's similar to what other backup services do.

Note that restic has more ways of specifying what should be ignored (case-insensitive matches and sentinel-file-based method), but:
 - this one seems most useful,
 - the other ones can be used manually,
 - the other ones seem to be useful in a pretty narrow set of cases between "things that can be covered by this one" and "things that require a custom paths provider".

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by robryk** [comment]
@bbigras  @i077 as maintainers of nixos/tests/restic (and, I assume, nixos/services/backups/restic.nix, though that file doesn't mention maintainers in any form)

**Review by bachp — APPROVED** [review]

**Review by i077 — APPROVED** [review]

**Comment by robryk** [comment]
Please do not merge this yet. The excludes are not appropriately escaped.

**Comment by robryk** [comment]
I've:
 - fulfilled the former TODO for actually testing the behaviour (by adding a restore test),
 - used `--exclude-file` instead of `--exclude` to Once And For All get rid of all potential shell escaping problems in exclude patterns.

**Review by i077 — APPROVED** [review]
LGTM, using the --exclude-file option is more robust than supplying each line as a CLI arg. You might also consider adding an `excludeFile` option to let users supply their own file (or maybe just keep this in `extraBackupArgs`).

**Comment by robryk** [comment]
> or maybe just keep this in extraBackupArgs

This is what I intend to do, given that (a) it's mostly trivial to do (b) other backup services do not expose such an option, so it's not like people expect to be able to do that without knowing about the commandline option.

Semi-relatedly: I have a bunch of changes on top of this that I'm still appending to (featuring currently things like making `paths` and paths-from-script be able to coexist, providing restic wrappers for easier restoring; featuring in the future: support for mirroring backups being created to additional repositories using copy). I intend to keep making a new PR with everything ready that I have at the moment the previous PR is merged. Please let me know if you prefer me to do that in some other way.

**Comment by i077** [comment]
Sounds good to me. Personally I'd prefer one "feature's" worth of changes at a time as that would make each one easier to review. That said, I like your ideas for this module. Thanks!

File: nixos/modules/services/backup/restic.nix
Comment (SuperSandro2000) [inline nixos/modules/services/backup/restic.nix:291]:
```suggestion
      message = "services.restic.backups.${n}: exactly one of repository or repositoryFile should be set";
```

Comment (robryk) [reply nixos/modules/services/backup/restic.nix]: Done.

**Comment by robryk** [comment]
Ping?

**Comment by NickCao** [comment]
@ofborg test restic

**Review by NickCao — APPROVED** [review]

**Pull request merged**
